src/multipart/multipart.rs

```diff
@@ -0,0 +1,469 @@
+extern crate uuid;
+
+use std;
+use std::borrow::Cow;
+use std::io::Read;
+use std::ascii::AsciiExt;
+use hyper::mime::Mime;
+
+/// A multipart/form-data request.
+#[derive(Debug)]
+pub struct MultipartRequest {
+    boundary: String,
+    fields: Vec<MultipartField>,
+}
+
+impl MultipartRequest {
+    /// Creates a new MultipartRequest without any content.
+    pub fn new() -> MultipartRequest {
+        MultipartRequest {
+            boundary: format!("{}", uuid::Uuid::new_v4().simple()),
+            fields: Vec::new(),
+        }
+    }
+    /// Adds a field, builder style.
+    pub fn field(mut self, field: MultipartField) -> MultipartRequest {
+        self.fields.push(field);
+        self
+    }
+    /// Adds multiple fields.
+    pub fn fields(&mut self, mut fields: Vec<MultipartField>) {
+        self.fields.append(&mut fields);
+    }
+    /// Turns this MultipartRequest into a RequestReader which implements the Read trait.
+    pub fn reader(self) -> RequestReader {
+        RequestReader::new(self)
+    }
+    /// Gets the automatically chosen boundary.
+    pub fn boundary(&self) -> &str {
+        &self.boundary
+    }
+    /// If predictable, computes the length the request will have
+    /// The length should be preditable if only String and file fields have been added,
+    /// but not if a generic reader has been added;
+    pub fn compute_length(&self) -> Option<u64> {
```

Comment:
Like `MultipartRequest::reader` this was pub so it could be used in other modules. I will turn it into a free public function in the module instead as in `reader`.